pkg/controlplane/controller/leaderelection/leaderelection_controller.go

```
@@ -0,0 +1,633 @@
+				lease.Spec.AcquireTime = nil
+				_, err = c.leaseClient.Leases(leaderLeaseID.namespace).Update(ctx, lease, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+			} else if lease.Spec.HolderIdentity != nil && *lease.Spec.HolderIdentity != electeeName {
+				klog.Infof("lease %q %q already exists for holder %q but should be held by %q, marking end of term", leaderLeaseID.namespace, leaderLeaseID.name, *lease.Spec.HolderIdentity, electee.Name)
+				if lease.Annotations == nil {
+					lease.Annotations = make(map[string]string)
+				}
+				lease.Annotations[EndOfTermAnnotationName] = "true"
+				_, err = c.leaseClient.Leases(leaderLeaseID.namespace).Update(ctx, lease, metav1.UpdateOptions{})
+				if err != nil {
+					return err
+				}
+
+				// Hack: force reconciliation
+				key, err := controller.KeyFunc(lease)
+				if err != nil {
+					utilruntime.HandleError(fmt.Errorf("cannot get name of object %v: %w", lease, err))
+				}
+				c.leaseQueue.AddRateLimited(key)
+			}
+
+			return nil
+		}
+		return err
+	}
+	return nil
+}
+
+func pickBestLeader(candidates []*v1alpha1.LeaseCandidate) *v1alpha1.LeaseCandidate {
+	var electee *v1alpha1.LeaseCandidate
+	for _, c := range candidates {
+		if electee == nil || compare(electee, c) > 0 {
+			electee = c
+		}
+	}
+	if electee == nil {
+		klog.Infof("pickBestLeader: none found")
+	} else {
+		klog.Infof("pickBestLeader: %s %s", electee.Namespace, electee.Name)
+	}
+	return electee
+}
+
+func shouldReelect(candidates []*v1alpha1.LeaseCandidate, currentLeader *v1alpha1.LeaseCandidate) bool {
+	klog.Infof("shouldReelect for candidates: %+v", candidates)
+	pickedLeader := pickBestLeader(candidates)
+	if pickedLeader == nil {
+		return false
+	}
+	return compare(currentLeader, pickedLeader) > 0
+}
+
+func getCompatibilityVersion(l *v1alpha1.LeaseCandidate) semver.Version {
+	value := l.Spec.CompatibilityVersion
+	v, err := semver.ParseTolerant(value)
+	if err != nil {
```

we should validate this in the API so we guarantee we never get an error here

---

Am not convinced hiding the error behind a zero value is a good idea. Go prefers explicit error handling. Here in the helper I cannot understand whether zero value does the right thing. I might be surprised if I call it.